src/datomish/db.cljc

```diff
@@ -12,31 +12,312 @@
+   ;; Assuming require [clojure.tools.logging :as log]
+   (Thread/setDefaultUncaughtExceptionHandler
+     (reify Thread$UncaughtExceptionHandler
+       (uncaughtException [_ thread ex]
+         (println ex "Uncaught exception on" (.getName thread))))))
 
 (defprotocol IDB
   (query-context
     [db])
-  (close
+
+  (close-db
+    [db]
+    "Close this database. Returns a pair channel of [nil error].")
+
+  (schema
+    [db]
+    "Return the schema of this database.")
+
+  (idents
     [db]
-    "Close this database. Returns a pair channel of [nil error]."))
+    "Return the known idents of this database, as a map from keyword idents to entids.")
+
+  (current-tx
+    [db]
+    "TODO: document this interface.")
+
+  (<eavt
+    [db pattern]
+    "Search for datoms using the EAVT index.")
+
+  (<avet
+    [db pattern]
+    "Search for datoms using the AVET index.")
+
+  (<apply-datoms
+    [db datoms]
+    "Apply datoms to the store.")
+
+  (<advance-tx
+    [db]
+    "TODO: document this interface."))
+
+;; TODO: handle _?
+(defn search->sql-clause [pattern]
+  (merge
+    {:select [:*] ;; e :a :v :tx] ;; TODO: generalize columns.
+     :from [:datoms]}
+    (if-not (empty? pattern)
+      {:where (cons :and (map #(vector := %1 (if (keyword? %2) (str %2) %2)) [:e :a :v :tx] pattern))} ;; TODO: use schema to process v.
+      {})))
 
-(defrecord DB [sqlite-connection]
+(defrecord DB [sqlite-connection schema idents current-tx]
+  ;; idents is map {ident -> entid} of known idents.  See http://docs.datomic.com/identity.html#idents.
   IDB
   (query-context [db] (context/->Context (source/datoms-source db) nil nil))
-  (close [db] (s/close (.-sqlite-connection db))))
 
-(defn <with-sqlite-connection [sqlite-connection]
+  (schema [db] (.-schema db))
+
+  (idents [db] (.-idents db))
+
+  (current-tx
+    [db]
+    (inc (:current-tx db)))
+
+  ;; TODO: use q for searching?  Have q use this for searching for a single pattern?
+  (<eavt [db pattern]
+    (go-pair
+      ;; TODO: find a better expression of this pattern.
+      (let [rows (<? (->>
+                       (search->sql-clause pattern)
+                       (sql/format)
+                       (s/all-rows (:sqlite-connection db))))]
+        (mapv #(Datom. (:e %) (:a %) (:v %) (:tx %) true) rows)))) ;; TODO: map values according to schema.
+
+  (<avet [db pattern]
+    (go-pair
+      ;; TODO: find a better expression of this pattern.
+      (let [[a v] pattern
+            rows (<? (->>
+                       {:select [:*] :from [:datoms] :where [:and [:= :a a] [:= :v v]]}
+                       (sql/format)
+                       (s/all-rows (:sqlite-connection db))))]
+        (mapv #(Datom. (:e %) (:a %) (:v %) (:tx %) true) rows)))) ;; TODO: map values according to schema.
+
+  (<apply-datoms [db datoms]
+    (go-pair
+      (let [exec (partial s/execute! (:sqlite-connection db))]
+        ;; TODO: batch insert, batch delete.
```

Comment:
I'm somewhere in the middle, which is an awkward place to be :(

At the moment, `DB` has a direct handle to the SQLite connection, and doesn't think about transactions.  `<transact!` is two-phase; it first does a bunch of reads (via DB), and then bulk applies updates-and-delete.  I think `<apply-datoms` might move to `IConnection`, but I don't know the final form for "stale" DBs, where the underlying store has moved on.